sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PredicateReorder.scala

```diff
@@ -0,0 +1,101 @@
+
+  // Return 1.0D - 1.0D / Int.MaxValue if the expression is not supported calculate
+  // filter selectivity to avoid rankingAnd always return 0.0D.
+  private def selectivity(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    filterEstimation.calculateFilterSelectivity(exp, false)
+      .getOrElse(1.0D - 1.0D / Int.MaxValue)
+  }
+
+  // Formula: (selectivity - 1.0D) / expression cost
+  private def rankingAnd(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    (selectivity(exp, filterEstimation) - 1.0D) / expressionCost(exp)
+  }
+
+  // Formula: (-selectivity) / expression cost
+  private def rankingOr(exp: Expression, filterEstimation: FilterEstimation): Double = {
+    -selectivity(exp, filterEstimation) / expressionCost(exp)
+  }
+
+  // The cost of a call expression e is computed as:
+  //   cost(exp) = typeSize + functionCost + cost(children).
+  private def expressionCost(exp: Expression): Double = exp match {
+      case e: Expression if e.children.isEmpty =>
+        e.dataType.defaultSize
+      case e: IsNull =>
+        e.dataType.defaultSize + 1.0D + e.children.map(expressionCost).sum
+      case e: IsNotNull =>
+        e.dataType.defaultSize + 1.0D + e.children.map(expressionCost).sum
+      case e: IsNaN =>
+        e.dataType.defaultSize + 1.0D + e.children.map(expressionCost).sum
+      case e: Not =>
+        e.dataType.defaultSize + 1.0D + e.children.map(expressionCost).sum
+      case e: BinaryOperator =>
+        e.dataType.defaultSize + 1.0D + e.children.map(expressionCost).sum
+      case e: StringRegexExpression =>
+        e.dataType.defaultSize + 2.0D + e.children.map(expressionCost).sum
+      case e @ In(_, list) =>
+        e.dataType.defaultSize + 2.0D * (list.size - 1)
+      case e @ InSet(_, set) =>
+        e.dataType.defaultSize + 2.0D * (set.size - 1)
+      case e: MultiLikeBase =>
+        e.dataType.defaultSize + 2.0D * e.patterns.size
+      case e: Cast =>
+        8.0D + e.dataType.defaultSize + e.children.map(expressionCost).sum
+      case e =>
+        32.0D + e.children.map(expressionCost).sum
+    }
+
+  // We do not recursively sort all expressions for performance.
+  private def reorderPredicates(exp: Expression, filterEstimation: FilterEstimation): Expression = {
+    exp match {
+      case _: Or =>
+        splitDisjunctivePredicates(exp)
+          .map(e => (e, rankingOr(e, filterEstimation))).sortWith(_._2 < _._2).map(_._1)
+          .reduceLeft(Or)
+      case _: And =>
+        splitConjunctivePredicates(exp)
+          .map(e => (e, rankingAnd(e, filterEstimation))).sortWith(_._2 < _._2).map(_._1)
+          .reduceLeft(And)
+      case _ => exp
+    }
+  }
+
+  def apply(plan: LogicalPlan): LogicalPlan = {
+    if (conf.cboEnabled && conf.predicateReorder) {
+      plan transform {
+        case f @ Filter(cond, _) if cond.references.toSet.subsetOf(f.stats.attributeStats.keySet) =>
```

Comment:
shall we do the same thing for join condition? And conditions in `If`, `CaseWhen`?

Reply:
Could we do it step by step?

Reply:
Could we do it step by step?